runtime/near-vm-logic/src/logic.rs

```diff
@@ -361,52 +352,45 @@ impl<'a> VMLogic<'a> {
     ///
     /// For nul-terminated string:
     /// `read_memory_base * num_bytes / 2 + read_memory_byte * num_bytes + utf16_decoding_base + utf16_decoding_byte * num_bytes`
-    fn get_utf16_string(&mut self, len: u64, ptr: u64) -> Result<String> {
+    fn get_utf16_string(&mut self, mut len: u64, ptr: u64) -> Result<String> {
         self.gas_counter.pay_base(utf16_decoding_base)?;
-        let mut u16_buffer;
         let max_len =
             self.config.limit_config.max_total_log_length.saturating_sub(self.total_log_length);
-        if len != u64::MAX {
-            let input = self.memory.view(&mut self.gas_counter, MemSlice { ptr, len })?;
-            if len % 2 != 0 {
-                return Err(HostError::BadUTF16.into());
+
+        let mem_view = if len == u64::MAX {
+            len = self.get_nul_terminated_utf16_len(ptr, max_len)?;
+            self.memory.view_for_free(MemSlice { ptr, len })
+        } else {
+            self.memory.view(&mut self.gas_counter, MemSlice { ptr, len })
+        }?;
+
+        let input = stdx::as_chunks_exact(&mem_view).map_err(|_| HostError::BadUTF16)?;
+        if len > max_len {
+            return self.total_log_length_exceeded(len);
+        }
+
+        self.gas_counter.pay_per(utf16_decoding_byte, len)?;
+        char::decode_utf16(input.into_iter().copied().map(u16::from_le_bytes))
+            .collect::<Result<String, _>>()
+            .map_err(|_| HostError::BadUTF16.into())
+    }
+
+    /// Helper function to get length of NUL-terminated UTF-16 formatted string
+    /// in guest memory.
+    ///
+    /// In other words, counts how many bytes are there to first pair of NUL
+    /// bytes.
+    fn get_nul_terminated_utf16_len(&mut self, ptr: u64, max_len: u64) -> Result<u64> {
+        let mut len = 0;
+        loop {
+            if self.memory.get_u16(&mut self.gas_counter, ptr + len)? == 0 {
```

Comment:
We are planning to enable `clippy::integer_arithmetic` checks. To make it obviously non panicky, I think saturating add would be better here.